src/wattsi.pas

```diff
@@ -2207,22 +2209,36 @@ procedure Save(const Document: TDocument; const FileName: AnsiString; const InSp
-                  else
-                  if (AnsiContainsStr(ClassValue, 'css')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/css?' + URLEncodedJSONContents)
-                  else
-                  if (AnsiContainsStr(ClassValue, 'js')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/js?' + URLEncodedJSONContents)
-                  else
-                  if (AnsiContainsStr(ClassValue, 'abnf')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/abnf?' + URLEncodedJSONContents)
-                  else
-                  if (AnsiContainsStr(ClassValue, 'html')) then
-                     HighlighterOutput := TFPCustomHTTPClient.SimpleGet(HighlightServerURL + '/html?' + URLEncodedJSONContents);
-                  HighlighterOutputByJSONContents[CurrentHighlightedElementJSON] := HighlighterOutput;
-                  Write(F, Trim(HighlighterOutput));
+                  try
+                     HTTPClient := TFPHTTPClient.Create(nil);
+                     Ss := TStringStream.Create('');
+                     if (AnsiContainsStr(ClassValue, 'idl')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/webidl?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'css')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/css?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'js')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/js?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'abnf')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/abnf?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'html')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/html?' + URLEncodedJSONContents, Ss, [200,400]);
+                     HighlighterOutput := Ss.Datastring;
+                     Ss.Free;
+                     if HTTPClient.ResponseStatusCode = 400 then
+                     begin
+                        Writeln('Error: Highlighter failed.');
+                        Write(HighlighterOutput);
+                        Halt(1);
+                     end;
+                     HighlighterOutputByJSONContents[CurrentHighlightedElementJSON] := HighlighterOutput;
+                     Write(F, Trim(HighlighterOutput));
+                  finally
+                     HTTPClient.Free;
+                  end;
                end;
             except
               on E: EHTTPClient do
```

Comment:
I guess maybe it'd happen if a non-expected status code came back.

Reply:
> I guess maybe it'd happen if a non-expected status code came back.

Right — but then the question becomes: What would cause a non-expected status code to come back? The only statuses the highlighter application code itself ever responds with are 200, 400, and 404. And it only responds with 404 if the input isn’t actually JSON, which won’t ever happen unless our code is just completely broken anyway. Outside of the highlighter application code — in the HTTP server library code — I guess there may be some cases where it responds with a 500 or other 5xx if there’s an internal failure before the highlighter application code ever gets hit. But if that happens, that’s some other bigger, more-obvious failure.

In short: I still don’t think that exception catch is ever likely to get reached in any normal cases :)